hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/HazelcastSqlCaseOperator.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlCase;
+import org.apache.calcite.sql.fun.SqlCaseOperator;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.InferTypes;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Custom Hazelcast {@link SqlCaseOperator} to override the default return type
+ * inference strategy for CASE.
+ *
+ * @see HazelcastSqlCase
+ */
+public final class HazelcastSqlCaseOperator extends SqlOperator {
+
+    public HazelcastSqlCaseOperator() {
+        super(SqlCaseOperator.INSTANCE.getName(), SqlKind.CASE, MDX_PRECEDENCE, true, null, InferTypes.RETURN_TYPE, null);
+    }
+
+    @Override
+    public void validateCall(SqlCall call, SqlValidator validator, SqlValidatorScope scope, SqlValidatorScope operandScope) {
+        SqlCaseOperator.INSTANCE.validateCall(call, validator, scope, operandScope);
+    }
+
+    @Override
+    public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+        // SqlCaseOperator is doing the same
+        return validateOperands(validator, scope, call);
+    }
+
+    @Override
+    public boolean checkOperandTypes(SqlCallBinding binding, boolean throwOnFailure) {
+        return SqlCaseOperator.INSTANCE.checkOperandTypes(binding, throwOnFailure);
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding binding) {
+        if (binding instanceof SqlCallBinding) {
+            SqlCallBinding sqlBinding = (SqlCallBinding) binding;
+            SqlCase call = (SqlCase) sqlBinding.getCall();
+            HazelcastSqlValidator validator = (HazelcastSqlValidator) sqlBinding.getValidator();
+
+            validator.getTypeCoercion().caseWhenCoercion(sqlBinding);
+            RelDataType caseType = validator.getKnownNodeType(call);
+            if (caseType == null) {
+                throw sqlBinding.newValidationError(RESOURCE.dynamicParamIllegal());
+            }
+
+            for (SqlNode thenOperand : call.getThenOperands()) {
+                RelDataType thenOperandType = validator.deriveType(sqlBinding.getScope(), thenOperand);
+                if (!HazelcastTypeSystem.canCast(thenOperandType, caseType)) {
+                    throw sqlBinding.newValidationError(RESOURCE.illegalMixingOfTypes());
```

Comment:
This exception also doesn't help the user much - which exact combination is illegal? We can add our custom error message here easily.

Reply:
case removed in c1a5855c8212e429b5d7245a1906713af01e624b and 7f29e69e24bcd44dec551b9d255171a48bdda0ec, https://github.com/hazelcast/hazelcast/issues/17329 created